### ISSUE TITLE
5.12.0-18.1-liquorix-amd64 with mg-LRU freezed almost 2 minutes, no OOM

### ISSUE THREAD

**Issue by hakavlad** [zen-kernel/zen-kernel#223]
Demo: https://www.youtube.com/watch?v=xNJACGvkv_s

Log: https://pastebin.com/cSrQX6vm

Freeze: 1:28 - 3:23

Without mg-LRU OOM killer comes during some seconds. With mg-LRU - during some minutes.

It's easy to reproduce also at my host system. 

During this stall file cache closed to minimum, PSI memory pressure closed to 100.

Same problem with 5.13-rc2-mg-LRU: https://github.com/hakavlad/cache-tests/tree/main/mg-LRU-v3_vs_classic-LRU/3-firefox-tail-OOM - this is repo with log files.

For example: PSI metrics during the stall: https://github.com/hakavlad/cache-tests/blob/main/mg-LRU-v3_vs_classic-LRU/3-firefox-tail-OOM/mg-LRU-1/psi2

Memory metrics: https://github.com/hakavlad/cache-tests/blob/main/mg-LRU-v3_vs_classic-LRU/3-firefox-tail-OOM/mg-LRU-1/mem2

For comparsion: logs without mg-LRU in such cases:
PSI: https://github.com/hakavlad/cache-tests/blob/main/mg-LRU-v3_vs_classic-LRU/3-firefox-tail-OOM/classic-LRU-1/psi2
Mem: https://github.com/hakavlad/cache-tests/blob/main/mg-LRU-v3_vs_classic-LRU/3-firefox-tail-OOM/classic-LRU-1/mem2
-- killer comes quickly and no such long stalls.

How to reproduce? Watch ytube and run `tail /dev/zero`. Expected behavior: OOM killer fixes the problem (this happened without mg-LRU). With mg-LRU: killer comes in some minutes, system freezes during some minutes.

**Comment by hakavlad** [comment]
@yuzhaogoogle  Could you comment this, please?

**Comment by hakavlad** [comment]
Yet another PSI metrics log during such stall: https://raw.githubusercontent.com/hakavlad/cache-tests/main/mg-LRU-v3_vs_classic-LRU/3-firefox-tail-OOM/mg-LRU-3/psi2

**Comment by hakavlad** [comment]
And yet another normal PSI metrics log in such cases: https://github.com/hakavlad/cache-tests/blob/main/mg-LRU-v3_vs_classic-LRU/3-firefox-tail-OOM/classic-LRU-2/psi2

**Comment by damentz** [comment]
@hakavlad 5.12.0-18.1 doesn't contain the mgLRU fixes, can you try the latest?  Or in other words, 5.12.0-18.1 is known bad and you need to upgrade.

Changelog below:
```
linux-liquorix (5.12-28) unstable; urgency=medium

  * add fix for multigenerational lru
   -  mm: fix priority queue in multigenerational lru
  * update version to 5.12.0-19.2

 -- Steven Barrett <steven@liquorix.net>  Wed, 21 Jul 2021 19:58:16 -0500

linux-liquorix (5.12-27) unstable; urgency=medium

  * merge 5.12.19
  * update version to 5.12.0-19.1

 -- Steven Barrett <steven@liquorix.net>  Tue, 20 Jul 2021 11:11:11 -0500

linux-liquorix (5.12-26) unstable; urgency=medium

  * merge 5.12.18
  * reset dirty ratios from 50%/20% (hard / background), to 20%/10%
  * merged le9 cache protection for mainline LRU
  * update version to 5.12.0-18.1

 -- Steven Barrett <steven@liquorix.net>  Mon, 19 Jul 2021 12:51:34 -0500
```

**Comment by yuzhaogoogle** [comment]
There are multiple aspects of this behavior.

From users' perspective, IMO, most users wouldn't want early oom kills. In this particular case, since "tail /dev/zero" serves no practical purpose, I'm not sure we could call it a valid use case. A boarder line use case for early oom kills is for users who open many apps and never bother to close those they don't use. On the other hand, early oom kills would be a nightmare for users who run large apps and are willing to wait for them to finish. Don't get me wrong -- I do agree users should be able to choose a behavior for themself.

From the technical perspective, kernel oom kills happen when page reclaim (LRU) fails. It doesn't take any PSI or user experience into account. Ironically, oom kills happen earlier with the mainline LRU is not by design. It's simply because the mainline LRU isn't good at finding pages to reclaim. Given kernel doesn't really know what users want, it makes decisions that meet all the requirement in terms of oom kills. So instead, it tries to hold on as long as it can.

Are there any solutions for this problem? The answer is yes. A good one is oomd, which monitors PSI and initiates kills on behalf of users. It's also possible to use memory cgroup to limit how much memory a system can overcommit. There are also vm.overcommit_*, if memory cgroup is not enabled.

I also could add a new option to MGLRU, and when set, page reclaim won't try as hard, which in turn will trigger oom kills early. But it won't be able to anticipate what users really want. It'd be more like a failsafe, in case of the aforementioned solutions are not in place.

**Comment by heftig** [comment]
I agree with this. If you want to avoid trashing you should be running a userspace oom watchdog that watches swap space and/or PSI and kills memory hogs before the system bogs down.

**Comment by hakavlad** [comment]
There are huge fantastic changes in .19.2. I'll report later.

**Comment by hakavlad** [comment]
Killer comes with File > 100 MB, no hangs. Killer comes EARLY. Cannot exhaust cache. This is the other extreme, the opposite of what it was before. But from the user's point of view, this behavior is better: the system remains manageable. I will continue testing, hopefully write more later.

**Comment by hakavlad** [comment]
```
июл 24 01:01:37 PC kernel: oom-kill:constraint=CONSTRAINT_NONE,nodemask=(null),cpuset=/,mems_allowed=0,global_oom,task_memcg=/system.slice/lightdm.service,task=Xorg,pid=4959,uid=0
июл 24 01:01:37 PC kernel: Out of memory: Killed process 4959 (Xorg) total-vm:395632kB, anon-rss:2188kB, file-rss:3968kB, shmem-rss:620kB, UID:0 pgtables:560kB oom_score_adj:0
июл 24 01:01:37 PC kernel: Purging GPU memory, 0 pages freed, 0 pages still pinned, 2112 pages left available.
июл 24 01:01:37 PC kernel: oom_reaper: reaped process 14441 (python3), now anon-rss:0kB, file-rss:0kB, shmem-rss:0kB
июл 24 01:01:37 PC kernel: nouveau 0000:01:00.0: Enabling HDA controller
июл 24 01:01:39 PC kernel: nouveau 0000:01:00.0: Enabling HDA controller
июл 24 01:01:21 PC polkitd(authority=local)[860]: Unregistered Authentication Agent for unix-session:4 (system bus name :1.29, object path /org/mate/PolicyKit1/AuthenticationAgent, locale ru_RU.UTF-8) (disconnec
июл 24 01:01:39 PC org.a11y.atspi.Registry[6314]: XIO:  fatal IO error 11 (Resource temporarily unavailable) on X server ":0"
июл 24 01:01:39 PC org.a11y.atspi.Registry[6314]:       after 763 requests (763 known processed) with 0 events remaining.
июл 24 01:01:20 PC unknown[6403]: mate-multiload-applet: Fatal IO error 11 (Ресурс временно недоступен) on X server :0.
июл 24 01:01:23 PC lightdm[6249]: pam_unix(lightdm:session): session closed for user user
июл 24 01:01:20 PC unknown[6398]: notification-area-applet: Fatal IO error 11 (Ресурс временно недоступен) on X server :0.
июл 24 01:01:20 PC unknown[6413]: mate-sensors-applet: Fatal IO error 11 (Ресурс временно недоступен) on X server :0.
июл 24 01:01:20 PC unknown[6405]: mate-brightness-applet: Fatal IO error 11 (Ресурс временно недоступен) on X server :0.
июл 24 01:01:20 PC unknown[6440]: clock-applet: Fatal IO error 11 (Ресурс временно недоступен) on X server :0.
июл 24 01:01:20 PC unknown[6392]: wnck-applet: Fatal IO error 11 (Ресурс временно недоступен) on X server :0.
июл 24 01:01:23 PC unknown[6515]: mate-system-monitor: Fatal IO error 11 (Ресурс временно недоступен) on X server :0.
июл 24 01:01:40 PC kernel: nouveau 0000:01:00.0: Enabling HDA controller
июл 24 01:01:43 PC lightdm[15422]: pam_unix(lightdm-greeter:session): session opened for user lightdm by (uid=0)
июл 24 01:01:43 PC systemd[1]: Created slice User Slice of lightdm.
июл 24 01:01:43 PC systemd[1]: lvm2-monitor.service: Cannot add dependency job, ignoring: Unit lvm2-monitor.service is masked.
июл 24 01:01:43 PC systemd[1]: lvm2-lvmpolld.socket: Cannot add dependency job, ignoring: Unit lvm2-lvmpolld.socket is masked.
июл 24 01:01:43 PC systemd[1]: lvm2-lvmetad.socket: Cannot add dependency job, ignoring: Unit lvm2-lvmetad.socket is masked.
июл 24 01:01:43 PC systemd[1]: Starting User Manager for UID 111...
июл 24 01:01:43 PC systemd-logind[749]: New session c2 of user lightdm.
июл 24 01:01:43 PC systemd[1]: Started Session c2 of user lightdm.
```
Just started memory hog. Xorg killer + fatal IO error. Maybe the problem is same as https://github.com/hakavlad/le9-patch/issues/5

**Comment by hakavlad** [comment]
So, the latest changes fixed the old problem and adds new problems.

**Comment by hakavlad** [comment]
Double kill:
```
июл 24 01:20:00 PC kernel: Out of memory: Killed process 31525 (tail) total-vm:15565224kB, anon-rss:8204456kB, file-rss:1396kB, shmem-rss:0kB, UID:1000 pgtables:30496kB oom_score_adj:0
июл 24 01:20:00 PC kernel: Purging GPU memory, 0 pages freed, 0 pages still pinned, 2112 pages left available.
июл 24 01:20:00 PC kernel: firefox-esr invoked oom-killer: gfp_mask=0x100cca(GFP_HIGHUSER_MOVABLE), order=0, oom_score_adj=0
июл 24 01:20:00 PC kernel: CPU: 1 PID: 20478 Comm: firefox-esr Not tainted 5.12.19-zen2 #1
июл 24 01:20:00 PC kernel: Hardware name: Acer VAG70_HC/VAG70, BIOS V2.16 01/14/2013
июл 24 01:20:00 PC kernel: Call Trace:
июл 24 01:20:00 PC kernel:  dump_stack+0x6e/0x8c
июл 24 01:20:00 PC kernel:  dump_header+0x4a/0x220
июл 24 01:20:00 PC kernel:  oom_kill_process+0xff/0x130
июл 24 01:20:00 PC kernel:  out_of_memory+0x10b/0x500
июл 24 01:20:00 PC kernel:  __alloc_pages_nodemask+0x1013/0x1140
июл 24 01:20:00 PC kernel:  pagecache_get_page+0x1ee/0x490
июл 24 01:20:00 PC kernel:  filemap_fault+0x7a0/0xc80
июл 24 01:20:00 PC kernel:  ? activate_page+0x186/0x240
июл 24 01:20:00 PC kernel:  ? do_set_pte+0xc3/0x130
июл 24 01:20:00 PC kernel:  ext4_filemap_fault+0x2c/0x40 [ext4]
июл 24 01:20:00 PC kernel:  __do_fault+0x4f/0x9a
июл 24 01:20:00 PC kernel:  handle_mm_fault+0x1044/0x1580
июл 24 01:20:00 PC kernel:  do_user_addr_fault+0x1a4/0x6e0
июл 24 01:20:00 PC kernel:  exc_page_fault+0x6b/0x150
июл 24 01:20:00 PC kernel:  ? asm_exc_page_fault+0x8/0x30
июл 24 01:20:00 PC kernel:  asm_exc_page_fault+0x1e/0x30
июл 24 01:20:01 PC kernel: RIP: 0033:0x7f10e0644610
июл 24 01:20:01 PC kernel: Code: Unable to access opcode bytes at RIP 0x7f10e06445e6.
июл 24 01:20:01 PC kernel: RSP: 002b:00007ffc71269838 EFLAGS: 00010246
июл 24 01:20:01 PC kernel: RAX: 00007f10e08f8920 RBX: 00007f10e3932ea0 RCX: 0000000000000001
июл 24 01:20:01 PC kernel: RDX: 0000000000000001 RSI: 0000000000000001 RDI: 00007f10b8314a00
июл 24 01:20:01 PC kernel: RBP: 000000007fffffff R08: 0000000000000033 R09: 0000000000000005
июл 24 01:20:01 PC kernel: R10: 00007f10a2b59a00 R11: 0000000000000000 R12: 00007ffc71269860
июл 24 01:20:01 PC kernel: R13: 00007ffc71269858 R14: 0000000000000000 R15: 00007f10e0644610
июл 24 01:20:01 PC kernel: Mem-Info:
июл 24 01:20:01 PC kernel: active_anon:1458712 inactive_anon:744325 isolated_anon:0
                               active_file:9714 inactive_file:365 isolated_file:32
                               unevictable:5066 dirty:0 writeback:0
                               slab_reclaimable:12124 slab_unreclaimable:21830
                               mapped:139635 shmem:130599 pagetables:18059 bounce:0
                               free:28079 free_pcp:135 free_cma:0
июл 24 01:20:01 PC kernel: Node 0 active_anon:5834848kB inactive_anon:2977300kB active_file:38856kB inactive_file:1460kB unevictable:20264kB isolated(anon):0kB isolated(file):128kB mapped:558540kB dirty:0kB writeback:0kB shmem:522396kB shmem_thp: 0kB shmem_pmdmapped: 0kB anon_thp: 0kB writeback_tmp:0kB kernel_stack:10272kB pagetables:72236kB all_unreclaimable? no
июл 24 01:20:01 PC kernel: Node 0 DMA free:15360kB min:104kB low:128kB high:152kB reserved_highatomic:0KB active_anon:0kB inactive_anon:0kB active_file:0kB inactive_file:0kB unevictable:0kB writepending:0kB present:15984kB managed:15360kB mlocked:0kB bounce:0kB free_pcp:0kB local_pcp:0kB free_cma:0kB
июл 24 01:20:01 PC kernel: lowmem_reserve[]: 0 1985 9440 9440 9440
июл 24 01:20:01 PC kernel: Node 0 DMA32 free:43784kB min:14188kB low:17732kB high:21276kB reserved_highatomic:0KB active_anon:2270688kB inactive_anon:4620kB active_file:7824kB inactive_file:600kB unevictable:0kB writepending:0kB present:2429692kB managed:2364156kB mlocked:0kB bounce:0kB free_pcp:0kB local_pcp:0kB free_cma:0kB
июл 24 01:20:01 PC kernel: lowmem_reserve[]: 0 0 7455 7455 7455
июл 24 01:20:01 PC kernel: Node 0 Normal free:53172kB min:53284kB low:66604kB high:79924kB reserved_highatomic:0KB active_anon:3563572kB inactive_anon:2973052kB active_file:30844kB inactive_file:988kB unevictable:20264kB writepending:0kB present:7854080kB managed:7637980kB mlocked:11808kB bounce:0kB free_pcp:540kB local_pcp:0kB free_cma:0kB
июл 24 01:20:01 PC kernel: lowmem_reserve[]: 0 0 0 0 0
июл 24 01:20:01 PC kernel: Node 0 DMA: 0*4kB 0*8kB 0*16kB 0*32kB 0*64kB 0*128kB 0*256kB 0*512kB 1*1024kB (U) 1*2048kB (M) 3*4096kB (M) = 15360kB
июл 24 01:20:01 PC kernel: Node 0 DMA32: 10*4kB (M) 12*8kB (UME) 24*16kB (UME) 13*32kB (UME) 4*64kB (UME) 7*128kB (UME) 3*256kB (UE) 24*512kB (UME) 5*1024kB (UM) 2*2048kB (UE) 5*4096kB (M) = 44840kB
июл 24 01:20:01 PC kernel: Node 0 Normal: 179*4kB (UME) 572*8kB (UME) 1061*16kB (UME) 390*32kB (UME) 110*64kB (UME) 15*128kB (UME) 5*256kB (ME) 17*512kB (ME) 0*1024kB 0*2048kB 0*4096kB = 53692kB
июл 24 01:20:01 PC kernel: Node 0 hugepages_total=0 hugepages_free=0 hugepages_surp=0 hugepages_size=2048kB
июл 24 01:20:01 PC kernel: 142144 total pagecache pages
июл 24 01:20:01 PC kernel: 14 pages in swap cache
июл 24 01:20:01 PC kernel: Swap cache stats: add 4696732, delete 4693569, find 605412/648457
июл 24 01:20:01 PC kernel: Free swap  = 30968108kB
июл 24 01:20:01 PC kernel: Total swap = 40069980kB
июл 24 01:20:01 PC kernel: 2574939 pages RAM
июл 24 01:20:01 PC kernel: 0 pages HighMem/MovableOnly
июл 24 01:20:01 PC kernel: 70565 pages reserved
июл 24 01:20:01 PC kernel: 0 pages hwpoisoned
июл 24 01:20:01 PC kernel: oom-kill:constraint=CONSTRAINT_NONE,nodemask=(null),cpuset=/,mems_allowed=0,global_oom,task_memcg=/user.slice/user-1000.slice/session-6.scope,task=Web Content,pid=13625,uid=1000
июл 24 01:20:01 PC kernel: Out of memory: Killed process 13625 (Web Content) total-vm:3616064kB, anon-rss:43772kB, file-rss:24904kB, shmem-rss:501464kB, UID:1000 pgtables:3204kB oom_score_adj:0
июл 24 01:20:01 PC kernel: oom_reaper: reaped process 31525 (tail), now anon-rss:0kB, file-rss:0kB, shmem-rss:0kB
```
`tail` killed and `Web Content` (innocent!) killed.

mem log: OOM even with huge swap free https://pastebin.com/eUAScbeR
```
2021-07-24 01:19:28,347: MA=208=2% MF=110 A=8866 F=306 AF=30 IF=276 D=0 C=306 SF=38898 SU=233
2021-07-24 01:19:28,848: MA=109=1% MF=110 A=8856 F=206 AF=27 IF=179 D=0 C=206 SF=38483 SU=648
2021-07-24 01:19:29,349: MA=28=0% MF=111 A=8869 F=85 AF=24 IF=61 D=0 C=85 SF=38118 SU=1013
2021-07-24 01:19:29,850: MA=17=0% MF=110 A=8790 F=66 AF=26 IF=39 D=0 C=66 SF=37718 SU=1413
2021-07-24 01:19:30,351: MA=8=0% MF=110 A=8710 F=48 AF=24 IF=24 D=0 C=48 SF=37243 SU=1888
2021-07-24 01:19:30,852: MA=7=0% MF=111 A=8705 F=45 AF=26 IF=19 D=0 C=45 SF=36661 SU=2470
2021-07-24 01:19:31,353: MA=6=0% MF=109 A=8697 F=44 AF=26 IF=18 D=0 C=44 SF=36069 SU=3062
2021-07-24 01:19:31,854: MA=9=0% MF=110 A=8687 F=47 AF=28 IF=19 D=0 C=47 SF=35535 SU=3596
2021-07-24 01:19:32,355: MA=10=0% MF=110 A=8675 F=48 AF=31 IF=17 D=0 C=48 SF=34946 SU=4184
2021-07-24 01:19:32,855: MA=12=0% MF=110 A=8661 F=51 AF=33 IF=18 D=0 C=51 SF=34364 SU=4767
2021-07-24 01:19:33,356: MA=14=0% MF=111 A=8655 F=51 AF=34 IF=16 D=0 C=51 SF=33793 SU=5338
2021-07-24 01:19:33,857: MA=14=0% MF=110 A=8638 F=52 AF=34 IF=18 D=0 C=52 SF=33220 SU=5910
2021-07-24 01:19:34,358: MA=18=0% MF=112 A=8627 F=55 AF=28 IF=27 D=0 C=55 SF=32665 SU=6466
2021-07-24 01:19:34,859: MA=18=0% MF=109 A=8615 F=58 AF=30 IF=28 D=0 C=58 SF=32068 SU=7063
2021-07-24 01:19:35,360: MA=19=0% MF=110 A=8604 F=60 AF=34 IF=25 D=0 C=60 SF=31469 SU=7662
2021-07-24 01:19:35,860: MA=21=0% MF=109 A=8592 F=64 AF=38 IF=26 D=0 C=64 SF=30874 SU=8257
2021-07-24 01:19:36,361: MA=17=0% MF=112 A=8595 F=49 AF=37 IF=12 D=0 C=49 SF=30299 SU=8832 // OOM
2021-07-24 01:19:36,862: MA=173=2% MF=274 A=8563 F=42 AF=29 IF=14 D=0 C=42 SF=35056 SU=4075
2021-07-24 01:19:37,363: MA=8243=84% MF=8345 A=553 F=48 AF=32 IF=16 D=0 C=48 SF=37639 SU=1492
2021-07-24 01:19:37,864: MA=8235=84% MF=8335 A=560 F=51 AF=35 IF=16 D=0 C=51 SF=37646 SU=1485
```
@damentz IMHO it would be a good idea to release a hotfix with MGLRU=0 (for lqx) until the bugs are fixed.

**Comment by yuzhaogoogle** [comment]
> mem log: OOM even with huge swap free https://pastebin.com/eUAScbeR

It doesn't matter how much free swap space there is when the disk it swaps to has zero bandwidth left.

Also oom kills don't happen at a constant pace -- I'd recommend running the test many times to get a reliable reading to ascertain whether this is a new problem. It'd be great if you could also post the mainline test results for comparison. If it can swap as much, I'd be surprised if it doesn't occasionally kill more than one app.

Why? Because sometimes oom kills don't immediately free memory and leaves it to oom reaper who cleans up after defunct processes. More memory in swap, the longer it takes for the entire oom procedure to finish. Meanwhile, oom killer could be victimizing more processes.

**Comment by hakavlad** [comment]
It was 2 attempts, and I got 2 double-kills:
-  https://github.com/zen-kernel/zen-kernel/issues/223#issuecomment-885742654
- https://github.com/zen-kernel/zen-kernel/issues/223#issuecomment-885771179

>It doesn't matter how much free swap space there is when the disk it swaps to has zero bandwidth left.

There is no such problem without mg-LRU: oom happens when SwapFree and cache echausted.

>when the disk it swaps to has zero bandwidth left

Disk has HIGH bandthwith. It is 8-stream zram.

>I'd recommend running the test many times to get a reliable reading to ascertain whether this is a new problem

OK, I'll continue.

**Comment by yuzhaogoogle** [comment]
Thanks.

I just took a closer look at your log.

The first oom kill happened:

> июл 24 01:20:00 PC kernel: Out of memory: Killed process 31525 (tail) total-vm:15565224kB, anon-rss:8204456kB, file-rss:1396kB, shmem-rss:0kB, UID:1000 pgtables:30496kB oom_score_adj:0

Then the second oom kill happened:

> июл 24 01:20:01 PC kernel: Out of memory: Killed process 13625 (Web Content) total-vm:3616064kB, anon-rss:43772kB, file-rss:24904kB, shmem-rss:501464kB, UID:1000 pgtables:3204kB oom_score_adj:0

Finally the oom reaper finished:

> июл 24 01:20:01 PC kernel: oom_reaper: reaped process 31525 (tail), now anon-rss:0kB, file-rss:0kB, shmem-rss:0kB

So it seems the problem is what I described: it took a little longer for the first oom procedure to free memory, meanwhile the system was still under memory pressure and had to trigger the second oom kill.

**Comment by yuzhaogoogle** [comment]
Also, as a side note:

It's better to have multiple small zram devices, which you already do, than one large device, because they help increase concurrency during swap.

Also it's important to have all devices at the same priority when swapping on. Otherwise, they don't help as much.

**Comment by damentz** [comment]
Adding some context here, Liquorix had PSI disabled in previous versions, so userspace OOM daemons couldn't manage memory pressure properly without information from PSI.  That's now resolved and built as part of https://github.com/zen-kernel/zen-kernel/issues/222, but `psi=1` still needs to be passed in to turn it on (for now).

Also, @yuzhaogoogle posted in https://github.com/zen-kernel/zen-kernel/issues/216#issuecomment-885869019 that there's room for improvement on the solution to swapping with mgLRU.  The swapping itself is a symptom of unexpected or unhandled memory pressure, so once that's resolved we'll probably get much better results.

**Comment by hakavlad** [comment]
@yuzhaogoogle Thanks for answers! 

>most users wouldn't want early oom kills

"early" is when progress is going on even with GUI frozen. 
In this case (with 5.12.0-18.1-liquorix-amd64) userspace is dead, payload is not going on, the system is unmaintainable. 
There is no reason to allow this behavior. 

>"tail /dev/zero" serves no practical purpose

Any process with an uncontrolled memory leak can be in place of the tail. 

>early oom kills would be a nightmare for users who run large apps and are willing to wait for them to finish

This is what I get with `5.12.19-zen2`. 

>I also could add a new option to MGLRU to limit how hard it should try before giving up.

This would be interesting.

**Comment by hakavlad** [comment]
So, the original problem is solved. Let's continue to discuss the new issue: https://github.com/zen-kernel/zen-kernel/issues/225.

**Comment by hakavlad** [comment]
v5.12.19-lqx4, the problem is back.

**Comment by hakavlad** [comment]
5.13.8-zen1, the problem still persists.

**Comment by hakavlad** [comment]
with swappiness=0: no progress, no OOM, system hangs:
```
$ mem2log
Starting mem2log with interval 2s, mode: 1
Process memory locked with MCL_CURRENT | MCL_FUTURE | MCL_ONFAULT
All values are in mebibytes
MemTotal: 9782.7, SwapTotal: 58696.3
--
MA is MemAvailable, MF is MemFree, A is Anon
F is File, AF is Active(file), IF is Inactive(file)
D is Dirty, C is Clean file (File - Dirty)
SF is SwapFree, SU is SwapUsed (SwapTotal - SwapFree)
--
MA=7367=75% MF=7424 A=1653 F=147 AF=40 IF=106 D=0 C=146 SF=58696 SU=0
MA=7360=75% MF=7417 A=1657 F=148 AF=40 IF=108 D=0 C=148 SF=58696 SU=0
MA=7361=75% MF=7416 A=1657 F=150 AF=41 IF=109 D=0 C=149 SF=58696 SU=0
MA=7359=75% MF=7415 A=1657 F=150 AF=41 IF=109 D=0 C=149 SF=58696 SU=0
MA=3915=40% MF=3971 A=5095 F=150 AF=41 IF=109 D=0 C=149 SF=58696 SU=0
MA=33=0% MF=111 A=9060 F=105 AF=13 IF=91 D=0 C=105 SF=58696 SU=0
MA=0=0% MF=110 A=9173 F=4 AF=3 IF=0 D=0 C=4 SF=58696 SU=0
MA=0=0% MF=110 A=9174 F=1 AF=1 IF=0 D=0 C=1 SF=58696 SU=0
MA=0=0% MF=111 A=9174 F=1 AF=1 IF=0 D=0 C=1 SF=58696 SU=0
MA=0=0% MF=110 A=9176 F=1 AF=1 IF=0 D=0 C=1 SF=58696 SU=0
MA=0=0% MF=110 A=9176 F=1 AF=1 IF=0 D=0 C=1 SF=58696 SU=0
MA=0=0% MF=110 A=9176 F=1 AF=1 IF=0 D=0 C=1 SF=58696 SU=0
MA=0=0% MF=111 A=9176 F=1 AF=1 IF=0 D=0 C=1 SF=58696 SU=0
MA=0=0% MF=110 A=9176 F=1 AF=1 IF=0 D=0 C=1 SF=58696 SU=0
MA=0=0% MF=111 A=9176 F=1 AF=1 IF=0 D=0 C=1 SF=58696 SU=0
MA=0=0% MF=111 A=9176 F=1 AF=0 IF=0 D=0 C=1 SF=58696 SU=0
MA=0=0% MF=110 A=9176 F=1 AF=1 IF=0 D=0 C=1 SF=58696 SU=0
MA=0=0% MF=111 A=9176 F=0 AF=0 IF=0 D=0 C=0 SF=58696 SU=0
MA=0=0% MF=111 A=9176 F=1 AF=0 IF=0 D=0 C=1 SF=58696 SU=0
MA=0=0% MF=110 A=9176 F=2 AF=2 IF=0 D=0 C=2 SF=58696 SU=0
MA=0=0% MF=111 A=9177 F=1 AF=1 IF=0 D=0 C=1 SF=58696 SU=0
MA=0=0% MF=110 A=9177 F=1 AF=1 IF=1 D=0 C=1 SF=58696 SU=0
MA=0=0% MF=111 A=9177 F=1 AF=1 IF=0 D=0 C=1 SF=58696 SU=0
MA=0=0% MF=111 A=9177 F=1 AF=0 IF=0 D=0 C=1 SF=58696 SU=0
MA=0=0% MF=110 A=9177 F=1 AF=1 IF=0 D=0 C=1 SF=58696 SU=0
MA=0=0% MF=110 A=9177 F=1 AF=1 IF=0 D=0 C=1 SF=58696 SU=0
MA=0=0% MF=110 A=9177 F=1 AF=1 IF=0 D=0 C=1 SF=58696 SU=0
MA=0=0% MF=110 A=9177 F=1 AF=1 IF=0 D=0 C=1 SF=58696 SU=0
MA=0=0% MF=110 A=9177 F=1 AF=0 IF=0 D=0 C=1 SF=58696 SU=0
MA=0=0% MF=110 A=9177 F=1 AF=0 IF=0 D=0 C=1 SF=58696 SU=0
MA=0=0% MF=109 A=9177 F=1 AF=1 IF=0 D=0 C=1 SF=58696 SU=0
MA=0=0% MF=110 A=9177 F=1 AF=0 IF=0 D=0 C=1 SF=58696 SU=0
MA=0=0% MF=111 A=9177 F=0 AF=0 IF=0 D=0 C=0 SF=58696 SU=0
MA=0=0% MF=110 A=9177 F=0 AF=0 IF=0 D=0 C=0 SF=58696 SU=0
MA=0=0% MF=110 A=9177 F=1 AF=1 IF=0 D=0 C=1 SF=58696 SU=0
MA=0=0% MF=110 A=9177 F=1 AF=0 IF=0 D=0 C=1 SF=58696 SU=0
MA=0=0% MF=111 A=9177 F=0 AF=0 IF=0 D=0 C=0 SF=58696 SU=0
MA=0=0% MF=110 A=9177 F=1 AF=1 IF=0 D=0 C=1 SF=58696 SU=0
MA=0=0% MF=111 A=9177 F=1 AF=1 IF=0 D=0 C=1 SF=58696 SU=0
MA=0=0% MF=110 A=9177 F=1 AF=1 IF=0 D=0 C=1 SF=58696 SU=0
MA=0=0% MF=110 A=9177 F=1 AF=1 IF=0 D=0 C=1 SF=58696 SU=0
MA=0=0% MF=110 A=9177 F=1 AF=0 IF=0 D=0 C=1 SF=58696 SU=0
MA=0=0% MF=109 A=9177 F=1 AF=1 IF=0 D=0 C=1 SF=58696 SU=0
MA=0=0% MF=110 A=9177 F=1 AF=1 IF=0 D=0 C=1 SF=58696 SU=0
MA=0=0% MF=110 A=9177 F=1 AF=0 IF=0 D=0 C=1 SF=58696 SU=0
MA=0=0% MF=110 A=9177 F=1 AF=0 IF=0 D=0 C=1 SF=58696 SU=0
MA=0=0% MF=110 A=9177 F=1 AF=1 IF=0 D=0 C=1 SF=58696 SU=0
MA=0=0% MF=110 A=9177 F=1 AF=1 IF=0 D=0 C=1 SF=58696 SU=0
MA=0=0% MF=110 A=9177 F=1 AF=1 IF=0 D=0 C=1 SF=58696 SU=0
MA=0=0% MF=110 A=9177 F=1 AF=0 IF=0 D=0 C=1 SF=58696 SU=0
MA=0=0% MF=110 A=9177 F=1 AF=1 IF=0 D=0 C=1 SF=58696 SU=0
MA=0=0% MF=111 A=9177 F=1 AF=1 IF=0 D=0 C=1 SF=58696 SU=0
MA=0=0% MF=111 A=9177 F=1 AF=0 IF=0 D=0 C=1 SF=58696 SU=0
MA=0=0% MF=110 A=9177 F=1 AF=0 IF=0 D=0 C=1 SF=58696 SU=0
MA=0=0% MF=110 A=9177 F=1 AF=1 IF=0 D=0 C=1 SF=58696 SU=0
MA=0=0% MF=111 A=9177 F=1 AF=0 IF=0 D=0 C=1 SF=58696 SU=0
MA=0=0% MF=110 A=9177 F=1 AF=1 IF=0 D=0 C=1 SF=58696 SU=0
MA=0=0% MF=110 A=9177 F=1 AF=0 IF=0 D=0 C=1 SF=58696 SU=0
MA=0=0% MF=110 A=9177 F=1 AF=1 IF=0 D=0 C=1 SF=58696 SU=0
MA=0=0% MF=110 A=9177 F=1 AF=1 IF=0 D=0 C=1 SF=58696 SU=0
MA=0=0% MF=110 A=9177 F=1 AF=0 IF=0 D=0 C=1 SF=58696 SU=0
MA=0=0% MF=110 A=9177 F=1 AF=1 IF=0 D=0 C=1 SF=58696 SU=0
MA=0=0% MF=110 A=9177 F=1 AF=1 IF=0 D=0 C=1 SF=58696 SU=0
MA=0=0% MF=110 A=9177 F=1 AF=0 IF=0 D=0 C=1 SF=58696 SU=0
MA=0=0% MF=110 A=9178 F=1 AF=1 IF=0 D=0 C=1 SF=58696 SU=0
```

**Comment by hakavlad** [comment]
@yuzhaogoogle 
Mainline lru has the following code: 
```
	/* The file pages on the current node are dangerously low */
	unsigned int file_is_tiny:1;
```
```
	/*
	 * If the system is almost out of file pages, force-scan anon.
	 */
	if (sc->file_is_tiny) {
		scan_balance = SCAN_ANON;
		goto out;
	}
```
And file cache may be protected by increasing vm.watermark_scale_factor, and the swapping goes with no problem with high vm.watermark_scale_factor even with `swappiness=0`.

With mgLRU: no OOM, no progress even with `watermark_scale_factor=1000`, cache may be closed to 0 during this stall.

Why mgLRU has no code such as `sc->file_is_tiny`?

**Comment by yuzhaogoogle** [comment]
"tiny" is a relative term. For tail /dev/zero, yes, it works. But in reality, most applications can't use this relative term to measure thrashing, e.g., databases whose majority of memory usage is file cache.

I'll propose something later today.

**Comment by hakavlad** [comment]
>"tiny" is a relative term

I mean that with ML-LRU it depends on `vm.watermark_scale_factor`, and it described in the doc:
>At 0, the kernel will not initiate swap until the amount of free and file-backed pages is less than the high watermark in a zone.

https://github.com/torvalds/linux/blob/v5.13/Documentation/admin-guide/sysctl/vm.rst#swappiness

i.e. with mainline LRU we can get cache protection and successful swapping by increasing `vm.watermark_scale_factor`. This does not happen with MGLRU:  system hangs with `vm.swappiness=0` with any `vm.watermark_scale_factor` (and this behavior is not documented for mgLRU). This in an another issue that might be reported.

We can get such behavior with ML-LRU after removing this code:
```
	/*
	 * If the system is almost out of file pages, force-scan anon.
	 */
	if (sc->file_is_tiny) {
		scan_balance = SCAN_ANON;
		goto out;
	}
```

**Comment by yuzhaogoogle** [comment]
You are right about sc->file_is_tiny: it prevents tail /dev/zero from exhaust file cache. But it doesn't prevent swap thrashing -- it hardly prevents file cache thrashing with test cases using higher amount of file cache.

vm.swappiness or vm.watermark_scale_factor shift the balance between swapping and dropping file cache, which might give the illusion that they protect file cache. Neither really fixes thrashing.

The problem with thrashing is new pages push out existing hot pages at an unacceptable rate. The problem with sc->file_is_tiny uses the concept of "size" rather than "rate". That's why MGLRU doesn't want it. I'm testing a patch that basically protects pages that are younger than N milliseconds. IMO, this is a lot better solution and it's more intuitive for users to configure.

**Comment by hakavlad** [comment]
>give the illusion that they protect file cache. Neither really fixes thrashing.

It's not issusion: cache is protected, and IO memory pressure closed to 0, and cache-bound operations work fast.

**Comment by yuzhaogoogle** [comment]
Let me give an example: create a file larger than memory, mmap it and read the mapped memory.

This is the counterpart to tail /dev/zero, i.e., stressing file cache instead of swapping.

**Comment by hakavlad** [comment]
I performed[1] some tests, but full results is not yet fully described and not yet reported.

In short, the results are the same as I previously[2] said:
- mgLRU cannot protect cache with swappiness=200 (It does not distinguish swappiness=2 and swappiness=200);
- mgLRU cannot protect cache with high watermark_scale_factor;
- freeze and no OOM (this issue).

I waiting for mgLRU v4 and hope to repeat the tests.

[1] https://github.com/hakavlad/cache-tests/tree/main/mg-LRU-v3_vs_classic-LRU
[2] https://forum.xanmod.org/thread-4102-post-7599.html#pid7599

**Comment by hakavlad** [comment]
>mmap it

mmap is a prerequisite?  Just read() is not OK?

**Comment by yuzhaogoogle** [comment]
Here you go:

#include <sys/stat.h>
#include <sys/mman.h> 
#include <errno.h>
#include <string.h>
#include <stdarg.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>

int main (int argc, char *argv[])
{
    int rc;
    int fd;
    struct stat stat;
    const char *ptr;
    int i;

    if (argc != 2)
      return -1;

    fd = open(argv[1], O_RDONLY);
    if (fd < 0) {
        printf("open %s failed: %s\n", argv[1], strerror(errno));
        return -1;
    }

    rc = fstat(fd, &stat);
    if (rc < 0) {
        printf("stat %s failed: %s\n", argv[1], strerror(errno));
        return -1;
    }

    ptr = mmap(0, stat.st_size, PROT_READ, MAP_PRIVATE, fd, 0);
    if (ptr == MAP_FAILED) {
        printf("mmap %s failed: %s\n", argv[1], strerror(errno));
        return -1;
    }

    for (i = 0; i < stat.st_size; i++) {
        if (!(i % (10 << 20)))
            printf("%x\n", ptr[i]);
    }

    return 0;
}

**Comment by hakavlad** [comment]
with ml-lru:
```
./bomb /media/user/4d0a188f-1e50-4c26-b0e1-4be70e082f73/user/f31.vdi
3c
0
0
ffffffff
0
ffffffff
ffffffc0
23
4
ffffffed
fffffff4
76
72
49                                                                                                                                                                                            
ffffff89                                                                                                                                                                                      
7f                                                                                                                                                                                            
ffffffde                                                                                                                                                                                      
68                                                                                                                                                                                            
0                                                                                                                                                                                             
7f                                                                                                                                                                                            
8                                                                                                                                                                                             
7f
ffffff80
fffffffd
7f
fffffffd
ffffff9c
7f
2f
ffffffb1
7f
2f
8
5f
7c
fffffffd
51
fffffffd
ffffff9f
7c
7
fffffffd
fffffffd
0
10
67
54
fffffffd
fffffffd
fffffffd
0
2f
2
0
69
ffffff9b
ffffff86
fffffff9
69
48
fffffffd
ffffff80
ffffff81
9
7f
47
ffffff8e
fffffff0
2a
a
3d
e
ffffffc1
0
69
1
ffffff90
ffffffc0
61
ffffff9e
55
ffffffae
0
4
2e
ffffffa0
55
ffffff80
ffffffff
75
0
ffffffa8
40
ffffffc0
55
65
ffffffa5
20
ffffffed
ffffffca
ffffffde
ffffffe8
ffffffd2
0
ffffffff
ffffffff
ffffff9c
ffffffe1
1
0
ffffffd9
ffffffd8
10
0
1
58
0
ffffff95
23
44
23
ffffffd0
fffffffd
3c
ffffffff
fffffffd
11
0
fffffffd
50
fffffffd
fffffffd
8
68
fffffff0
0
0
0
0
0
0
ffffffaa
4d
0
24
4
ffffffee
0
29
0
70
7
7f
1c
22
0
0
ffffffe5
2
7b
6e
ffffffcc
59
0
75
a
ffffff80
0
fffffff3
ffffffa2
67
ffffffc8
ffffffc4
46
5b
19
15
0
48
ffffffef
0
ffffff80
ffffff9d
19
4b
1
31
0
fffffff4
45
ffffffae
ffffffe9
70
16
ffffffac
ffffffa9
0
48
4
48
78
58
20
fffffff2
10
Segmentation error 
```
And what the result I should expect?

**Comment by hakavlad** [comment]
The same result with mglru.

**Comment by yuzhaogoogle** [comment]
Try a file a lot larger than your memory, e.g.,

stat -c %s test.file
97197188878

**Comment by yuzhaogoogle** [comment]
Anyway, I have the patch fix ready. It should give you the behavior you are looking for. I'll finish the test tomorrow and post it. Will keep you posted.

**Comment by hakavlad** [comment]
I tried file larger than memory. MemTotal: 10G, file: 22G.

OK, thanks.

**Comment by hakavlad** [comment]
This code works:
```
#!/usr/bin/env python3

from mmap import ACCESS_READ, mmap
from sys import argv

with open(argv[1], 'rb') as f:
    mm = mmap(f.fileno(), 0, access=ACCESS_READ)
    mm.read()
```
I've get early swapping with this code. Here we can open the new issue: overswapping even with swappiness=2 when reading mmapped file.

**Comment by yuzhaogoogle** [comment]
Can you please test the following cases on the mainline kernel and Zen kernel after #231 is released?

Test cases: 1) tail /dev/zero 2) your mmap python script

On the mainline kernel, no additional actions needed.
On Zen kernel, `echo 1000 > /sys/kernel/mm/lru_gen/min_ttl_ms`. Basically this protects memory used within 1000 milliseconds You can adjust this value and see how it affects things. I've tested 1000 and it works perfectly on my workstation while playing YouTube.

Thanks.

**Comment by hakavlad** [comment]
Min 100 ttls ms fixes the problem (freeze and no OOM with `tail /dev/zero`).

**Comment by hakavlad** [comment]
>2) your mmap python script

Nothing changed. Overswapping stayed.

**Comment by yuzhaogoogle** [comment]
1) On my workstation, 200 seems too small, probably because it has high memory (200G).

2) What do you mean by overswapping? Do you expect oom kills immediately with mmap? What's the mainline behavior?

Thanks.

**Comment by yuzhaogoogle** [comment]
If the original problem is fixed, please close this issue.

Please open a new issue, describe the behavior with the mainline kernel, the behavior with MGLRU, and the behavior you prefer to have. Attach the test case (the python script), the test results (like the video demo with tail /dev/zero), and the size of the file(s). (And it'd be great if you try different file sizes and read them more than once.)

Note that terms like "tiny" and "over" are relative or even subjective. We can interpret them differently and therefore have a hard time communicating with each other.

**Comment by hakavlad** [comment]
OK, thanks.

**Comment by hakavlad** [comment]
@yuzhaogoogle I'd like to test mgLRU v4 and continue our discussions at LKML. I think this is a broad topic that might be of interest to people at LKML.

**Comment by yuzhaogoogle** [comment]
That sounds a great idea. Thanks for participating.

Please shoot me an email. Once I have your email address, I'll cc you on the next version (v4) of MGLRU, which I'll be posting sometime next week.

**Comment by buzzcut-s** [comment]
> Also a side note:
> 
> It's better to have multiple small zram devices, which you already do, than one large device, because they help increase concurrency during swap.
> 
> Also it's important to have all devices at the same priority when swapping on. Otherwise, they don't help as much.

Sorry for bumping this old thread, but I was curious about this. From the kernel [docs](https://www.kernel.org/doc/html/latest/admin-guide/blockdev/zram.html) : 

> 2) Set max number of compression streams
>
> Regardless of the value passed to this attribute, ZRAM will always allocate multiple compression streams - one per online CPU - thus allowing several concurrent compression operations. The number of allocated compression streams goes down when some of the CPUs become offline. There is no single-compression-stream mode anymore, unless you are running a UP system or have only 1 CPU online.

Is it actually *still* beneficial to have multiple small zram devices at same priority vs one large device? zramctl shows that my zram device has as many comp streams as logical cpus (12) - I'd think multiple devices with 12 stream would actually cause more overhead & contention than just one device. 

Would really appreciate an answer especially since zram documentation and testing is both very lacking and very old.

**Comment by yuzhaogoogle** [comment]
1) Using multiple zram devices has nothing to do with compression streams
2) max_comp_streams is a NOP

Each swap device uses a single swap_info_struct->lock for something called swap cache. If there is only one zram device but there are many CPUs, this lock can be easily contented when swap is busy.

The last comment also reminds me of another one you made on memory.swap.max after I suggested it to a XanMod user -- it'd be more constructive if you fully understood what others are suggesting before dismissing them out of hand. Or better yet, ask why and let others explain their understandings, rather than make your own assumptions. Just my two cents.

**Comment by hakavlad** [comment]
@yuzhaogoogle Could you please answer this question:

>Is it actually still beneficial to have multiple small zram devices at same priority vs one large device?

(Sorry, I didn't see a direct answer.)

>reminds me of another one you made on memory.swap.max

>let others explain their understandings

I didn’t stop you from explaining your understanding, I just suggested the better solution. I have already tested your proposed solution it in practice and found it ineffective. The solution I proposed has shown its effectiveness in practice (I mean reducing `memory.high` instead of `memory.swap.max` to reduce file thrashing).

**Comment by yuzhaogoogle** [comment]
Yes, it is, if there are many CPUs and the system swaps heavily.

`swap_info_struct->lock` contention: https://lwn.net/Articles/704478/

This is an old article but it still largely applies.

**Comment by yuzhaogoogle** [comment]
Also it shouldn't be hard to verify the difference:

1) use one large zram device
2) use N zram devices, N is a half of the number of CPUs and the total size of these N zram devices is equal to that of 1). These devices should have the same priority.

And create a large tmpfs file, e.g, 2x of physical memory, and read it a few times, and see which one above is faster.

**Comment by hakavlad** [comment]
BTW, high min_ttl_ms value cause the same problem.

For example, system freeze with `min_ttl_ms=600000`
```
$ mem2log
Starting mem2log with interval 2s, mode: 1
Process memory locked with MCL_CURRENT | MCL_FUTURE | MCL_ONFAULT
All values are in mebibytes
MemTotal: 9782.7, SwapTotal: 0.0
--
MA is MemAvailable, MF is MemFree, A is Anon
F is File, AF is Active(file), IF is Inactive(file)
D is Dirty, C is Clean file (File - Dirty)
SF is SwapFree, SU is SwapUsed (SwapTotal - SwapFree)
--
...
MA=6241=64% MF=5933 A=2935 F=517 AF=8 IF=509 D=6 C=511 SF=9783 SU=0
MA=6240=64% MF=5931 A=2935 F=517 AF=8 IF=509 D=6 C=511 SF=9783 SU=0
MA=6234=64% MF=5926 A=2940 F=517 AF=8 IF=509 D=6 C=511 SF=9783 SU=0
MA=5183=53% MF=4874 A=3989 F=517 AF=8 IF=509 D=6 C=512 SF=9783 SU=0
MA=86=1% MF=110 A=9107 F=189 AF=17 IF=171 D=0 C=189 SF=9397 SU=386
MA=8=0% MF=110 A=9066 F=55 AF=40 IF=14 D=0 C=54 SF=8589 SU=1194
MA=12=0% MF=110 A=8866 F=62 AF=50 IF=12 D=0 C=62 SF=7768 SU=2014
MA=17=0% MF=109 A=8765 F=68 AF=64 IF=3 D=0 C=68 SF=5409 SU=4374
MA=25=0% MF=109 A=8765 F=75 AF=75 IF=0 D=0 C=75 SF=1746 SU=8036
MA=4=0% MF=110 A=8787 F=29 AF=24 IF=5 D=0 C=29 SF=4101 SU=5682
MA=12=0% MF=110 A=8733 F=45 AF=45 IF=0 D=0 C=45 SF=1923 SU=7860
MA=12=0% MF=111 A=8685 F=42 AF=40 IF=1 D=0 C=42 SF=180 SU=9602
MA=0=0% MF=110 A=8725 F=2 AF=2 IF=0 D=0 C=2 SF=0 SU=9783
MA=0=0% MF=109 A=8725 F=2 AF=2 IF=0 D=0 C=2 SF=0 SU=9783
MA=0=0% MF=110 A=8725 F=3 AF=2 IF=1 D=0 C=3 SF=0 SU=9783
MA=0=0% MF=111 A=8725 F=2 AF=1 IF=1 D=0 C=2 SF=0 SU=9783
MA=0=0% MF=111 A=8725 F=2 AF=2 IF=0 D=0 C=2 SF=0 SU=9783
MA=0=0% MF=110 A=8727 F=1 AF=1 IF=0 D=0 C=1 SF=0 SU=9783
MA=0=0% MF=110 A=8727 F=2 AF=2 IF=0 D=0 C=2 SF=0 SU=9783
MA=0=0% MF=110 A=8728 F=1 AF=1 IF=0 D=0 C=1 SF=0 SU=9783
MA=0=0% MF=110 A=8728 F=1 AF=1 IF=0 D=0 C=1 SF=0 SU=9783
MA=0=0% MF=110 A=8728 F=1 AF=1 IF=0 D=0 C=1 SF=0 SU=9783 // Freeze, low cache
MA=0=0% MF=110 A=8728 F=1 AF=1 IF=0 D=0 C=1 SF=0 SU=9783
MA=0=0% MF=110 A=8728 F=1 AF=1 IF=0 D=0 C=1 SF=0 SU=9783
MA=0=0% MF=109 A=8728 F=1 AF=1 IF=0 D=0 C=1 SF=0 SU=9783
MA=0=0% MF=112 A=8728 F=1 AF=1 IF=0 D=0 C=1 SF=0 SU=9783
MA=0=0% MF=111 A=8728 F=1 AF=1 IF=0 D=0 C=1 SF=0 SU=9783
MA=0=0% MF=111 A=8728 F=1 AF=1 IF=0 D=0 C=1 SF=0 SU=9783
MA=0=0% MF=111 A=8728 F=1 AF=1 IF=0 D=0 C=1 SF=0 SU=9783
MA=0=0% MF=111 A=8728 F=1 AF=1 IF=0 D=0 C=1 SF=0 SU=9783
MA=0=0% MF=111 A=8728 F=1 AF=1 IF=0 D=0 C=1 SF=0 SU=9783
MA=0=0% MF=110 A=8728 F=1 AF=1 IF=0 D=0 C=1 SF=0 SU=9783
MA=0=0% MF=110 A=8728 F=1 AF=1 IF=0 D=0 C=1 SF=0 SU=9783
MA=0=0% MF=111 A=8728 F=1 AF=1 IF=0 D=0 C=1 SF=0 SU=9783
MA=0=0% MF=110 A=8728 F=1 AF=1 IF=0 D=0 C=1 SF=0 SU=9783
MA=0=0% MF=111 A=8729 F=1 AF=1 IF=0 D=0 C=1 SF=0 SU=9783
MA=0=0% MF=110 A=8729 F=1 AF=1 IF=0 D=0 C=1 SF=0 SU=9783
MA=0=0% MF=110 A=8729 F=1 AF=1 IF=0 D=0 C=1 SF=0 SU=9783
MA=0=0% MF=110 A=8729 F=1 AF=1 IF=0 D=0 C=1 SF=0 SU=9783
MA=0=0% MF=110 A=8729 F=1 AF=1 IF=0 D=0 C=1 SF=0 SU=9783
MA=0=0% MF=110 A=8729 F=1 AF=1 IF=0 D=0 C=1 SF=0 SU=9783
MA=0=0% MF=110 A=8729 F=1 AF=1 IF=0 D=0 C=1 SF=0 SU=9783
MA=0=0% MF=111 A=8729 F=1 AF=1 IF=0 D=0 C=1 SF=0 SU=9783
MA=0=0% MF=110 A=8729 F=1 AF=1 IF=0 D=0 C=1 SF=0 SU=9783
MA=0=0% MF=110 A=8729 F=1 AF=1 IF=0 D=0 C=1 SF=0 SU=9783
MA=0=0% MF=110 A=8729 F=1 AF=1 IF=0 D=0 C=1 SF=0 SU=9783
MA=0=0% MF=110 A=8729 F=1 AF=0 IF=1 D=0 C=1 SF=0 SU=9783
MA=0=0% MF=111 A=8729 F=1 AF=1 IF=0 D=0 C=1 SF=0 SU=9783
MA=0=0% MF=110 A=8729 F=1 AF=1 IF=0 D=0 C=1 SF=0 SU=9783
MA=0=0% MF=110 A=8729 F=1 AF=1 IF=0 D=0 C=1 SF=0 SU=9783
MA=0=0% MF=110 A=8730 F=1 AF=0 IF=0 D=0 C=1 SF=0 SU=9783
MA=0=0% MF=110 A=8730 F=0 AF=0 IF=0 D=0 C=0 SF=0 SU=9783
MA=0=0% MF=110 A=8730 F=0 AF=0 IF=0 D=0 C=0 SF=0 SU=9783
MA=0=0% MF=115 A=8713 F=1 AF=0 IF=0 D=0 C=1 SF=204 SU=9579
MA=8450=86% MF=8562 A=281 F=23 AF=23 IF=0 D=0 C=23 SF=7120 SU=2663
MA=8438=86% MF=8541 A=285 F=41 AF=41 IF=0 D=0 C=40 SF=7124 SU=2659
MA=8422=86% MF=8516 A=293 F=59 AF=59 IF=0 D=0 C=59 SF=7132 SU=2651
MA=8322=85% MF=8405 A=387 F=81 AF=81 IF=0 D=0 C=81 SF=7233 SU=2550
MA=7992=82% MF=8069 A=755 F=97 AF=96 IF=0 D=0 C=96 SF=7645 SU=2137
MA=7951=81% MF=8021 A=799 F=107 AF=106 IF=1 D=0 C=107 SF=7691 SU=2092
MA=7863=80% MF=7927 A=862 F=120 AF=119 IF=2 D=0 C=120 SF=7751 SU=2031
MA=7782=80% MF=7844 A=943 F=128 AF=125 IF=3 D=1 C=127 SF=7830 SU=1952
MA=7768=79% MF=7823 A=954 F=139 AF=129 IF=10 D=5 C=134 SF=7839 SU=1943
MA=7759=79% MF=7813 A=973 F=144 AF=133 IF=11 D=5 C=139 SF=7859 SU=1923
MA=7745=79% MF=7796 A=991 F=147 AF=135 IF=12 D=5 C=142 SF=7868 SU=1915
MA=7733=79% MF=7783 A=1002 F=150 AF=137 IF=12 D=0 C=150 SF=7870 SU=1912
MA=7713=79% MF=7762 A=1023 F=153 AF=139 IF=13 D=0 C=152 SF=7875 SU=1907
MA=7696=79% MF=7743 A=1039 F=156 AF=142 IF=14 D=0 C=156 SF=7884 SU=1898
MA=7678=78% MF=7720 A=1053 F=166 AF=143 IF=22 D=3 C=162 SF=7891 SU=1892
```
